src/inx/client.rs

```diff
@@ -45,17 +62,84 @@ impl Inx {
+                commitment_slot: slot.0,
+                commitment_id: None,
+            })
+            .await?
+            .try_convert()
+    }
+
+    /// Get a stream of finalized slots.
+    pub async fn get_finalized_slots(
         &mut self,
         request: SlotRangeRequest,
     ) -> Result<impl Stream<Item = Result<Commitment, InxError>>, InxError> {
-        Ok(self
-            .inx
-            .listen_to_commitments(proto::SlotRangeRequest::from(request))
+        struct StreamState {
+            inx: Option<Inx>,
+            latest_finalized_slot: u32,
+            curr_slot: u32,
+            last_slot: u32,
+        }
+
+        let latest_finalized_slot = self
+            .get_node_status()
             .await?
-            .into_inner()
-            .map(|msg| msg?.try_convert()))
+            .latest_finalized_commitment
+            .commitment_id
+            .slot_index()
+            .0;
+        Ok(futures::stream::unfold(
+            StreamState {
+                inx: Some(self.clone()),
+                latest_finalized_slot,
+                curr_slot: request.start_slot(),
+                last_slot: request.end_slot(),
+            },
+            |mut state| async move {
+                // Inner function definition to simplify result type
+                async fn next(state: &mut StreamState) -> Result<Option<Commitment>, InxError> {
+                    let Some(inx) = state.inx.as_mut() else { return Ok(None) };
+
+                    if state.last_slot != 0 && state.curr_slot > state.last_slot {
+                        return Ok(None);
+                    }
+
+                    // If the current slot is not yet finalized, we will wait.
+                    if state.latest_finalized_slot < state.curr_slot {
+                        let mut status_changes = inx.listen_to_status_changes().await?;
+                        loop {
+                            match status_changes.try_next().await? {
+                                Some(status) => {
+                                    // If the status change updated the latest finalized commitment, we can continue.
+                                    if status.latest_finalized_commitment.commitment_id.slot_index().0
+                                        > state.latest_finalized_slot
+                                    {
+                                        state.latest_finalized_slot =
+                                            status.latest_finalized_commitment.commitment_id.slot_index().0;
+                                        break;
+                                    }
+                                }
+                                None => {
+                                    return Ok(None);
+                                }
+                            }
+                        }
+                    }
+                    let commitment = inx.get_committed_slot(state.curr_slot.into()).await?;
```

Comment:
not entirely sure I understand the call wotj this argument tbh ... if we assume that `state.latest_finalized_slot` got updated in the previouse loop, then it still could be `< state.curr_slot`, no? :thinking: shouldn't we call it with `state.latest_finalized_slot` instead? Or is `curr_slot` in the request always set so that can't be a problem?

Reply:
This iterates over the indexes in order, so we always get the slot with the current index. The loop above just waits until the status reports that the current index is finalized.

Reply:
it loops until it returns that the slot is finalized.